src/async.zsh

```diff
@@ -0,0 +1,53 @@
+
+#--------------------------------------------------------------------#
+# Async                                                              #
+#--------------------------------------------------------------------#
+
+_zsh_autosuggest_async_fetch_suggestion() {
+	local strategy_function="_zsh_autosuggest_strategy_$ZSH_AUTOSUGGEST_STRATEGY"
+	local prefix="$(_zsh_autosuggest_escape_command "$1")"
+
+	# Send the suggestion command to the pty to fetch a suggestion
+	zpty -w -n $ZSH_AUTOSUGGEST_PTY_NAME "$strategy_function '$prefix'"$'\0'
+}
+
+_zsh_autosuggest_async_suggestion_worker() {
+	local last_pid
+
+	while read -d $'\0' cmd; do
+		# Kill last bg process
+		kill -KILL $last_pid &>/dev/null
+
+		# Run suggestion search in the background
+		print -n -- "$(eval "$cmd")"$'\0' &
+
+		# Save the bg process's id so we can kill later
+		last_pid=$!
+	done
+}
+
+_zsh_autosuggest_async_suggestion_ready() {
+	# while zpty -rt $ZSH_AUTOSUGGEST_PTY_NAME suggestion 2>/dev/null; do
+	while read -u $_ZSH_AUTOSUGGEST_PTY_FD -d $'\0' suggestion; do
```

Comment:
Using `zpty -r` to read the FD here would allow supporting older versions of zsh e.g. through `kill`-signals. Doesn't support delimiter, but you can use a PATTERN as the last parameter, e.g. `zpty -rt $ZSH_AUTOSUGGEST_PTY_NAME suggestion '*'$'\0'`. I've not used pattern before, though, so I could be wrong on the syntax.